Add answers to multichoice, add method to resolve, a bunch more

LivInTheLookingGlass opened the pull request.

LivInTheLookingGlass committed:
Add answers field to multiple choice markets

LivInTheLookingGlass committed:
Consolidate from_dict()

LivInTheLookingGlass committed:
Add resolution  methods

LivInTheLookingGlass committed:
Correct inheritance issue

LivInTheLookingGlass committed:
Correct a typo

bcongdon commented on pymanifold/lib.py:
```diff
@@ -149,3 +163,54 @@ def _create_market(
+            market.id
+        except AttributeError:
+            market = self.get_market_by_id(market)
+        match market.outcomeType:
```
Neat, I haven't used match arms in Python yet. Do you know what the minimum Python version is?

LivInTheLookingGlass replied:
I actually rewrote that because I was running into versioning problems at the time

bcongdon commented on pymanifold/lib.py:
```diff
@@ -149,3 +163,54 @@ def _create_market(
         )
         return LiteMarket.from_dict(response.json())
+
+    def resolve_market(self, market: Union[LiteMarket, str], *args, **kwargs):
```
Can you add tests for these new functions?

bcongdon reviewed (commented):
Thanks for adding this!

LivInTheLookingGlass committed:
PseudoNumeric market fixes

LivInTheLookingGlass committed:
Prep for numeric API change

LivInTheLookingGlass committed:
Resolve some bugs found while testing numerics

LivInTheLookingGlass committed:
Add support for limit bets

LivInTheLookingGlass committed:
Add free resonse markets

LivInTheLookingGlass committed:
Add defaults, fix whitespace

LivInTheLookingGlass committed:
Compatibility with Python <3.10

LivInTheLookingGlass committed:
Fix all mypy, pydocstyle errors/warnings

LivInTheLookingGlass committed:
Hotfix for a bug in Manifold's API

LivInTheLookingGlass committed:
Refactor fetch API, add bets

LivInTheLookingGlass committed:
Add initial code for groups, users

LivInTheLookingGlass committed:
Merge from upstream

LivInTheLookingGlass committed:
Merge branch 'bcongdon-main'

LivInTheLookingGlass committed:
Update due to API changes

LivInTheLookingGlass committed:
Support multiple choice markets

LivInTheLookingGlass committed:
Add user s upport

LivInTheLookingGlass committed:
Pseudonumeric should no longer need two values to resolve

LivInTheLookingGlass committed:
Revert previous commit

LivInTheLookingGlass committed:
Add typing for DictDeserializable

LivInTheLookingGlass committed:
Improve typing a bit

LivInTheLookingGlass committed:
Improve pseudonumeric resolution

LivInTheLookingGlass committed:
Include some resolution details

LivInTheLookingGlass committed:
Reduce runtime cost of type hinting

LivInTheLookingGlass committed:
Let requests do exception raising

LivInTheLookingGlass committed:
Add from_url for convenience

LivInTheLookingGlass committed:
Now I can test this while it's a submodule

LivInTheLookingGlass committed:
Solve large majority of linter errors

LivInTheLookingGlass committed:
No more mypy errors

LivInTheLookingGlass commented:
Currently working on tests. Might take me slightly longer to get the ones here, as I'm very used to `pytest`. I assume a lot will translate

Additions to test:
- [ ] get_markets
- [ ] stream_markets
- [ ] list_bets
- [ ] get_bets
- [ ] stream_bets
- [ ] get_market_by_url
- [ ] get_user
- [ ] cancel_market
- [ ] create_multiple_choice_market
- [ ] _resolve_binary_market
- [ ] _resolve_pseudo_numeric_market
- [ ] _resolve_free_response_market
- [ ] _resolve_multiple_choice_market
- [ ] Group
- [ ] LiteUser

bcongdon commented:
Sounds great! Let me know when you'd like me to take a look

LivInTheLookingGlass committed:
Remove stream methods

LivInTheLookingGlass committed:
Correct typing, add slug

LivInTheLookingGlass committed:
Add tests for half of added/untested methods

LivInTheLookingGlass commented:
Okay, I've added tests for maybe half of my additions. Those are certainly ready to check out. I also removed two functions that I had added. They didn't have enough value prop for how hard they were to test. The following is ready for review, but if you'd rather wait until done, I totally get it

- [x] get_markets
- [x] list_groups
- [x] get_groups
- [x] get_group
- [x] list_bets
- [x] get_bets
- [x] get_market_by_url
- [x] get_user
- [x] cancel_market
- [x] create_multiple_choice_market
- [x] _resolve_binary_market
- [x] _resolve_pseudo_numeric_market
- [x] _resolve_free_response_market
- [x] _resolve_multiple_choice_market
- [x] Group
- [x] LiteUser

LivInTheLookingGlass committed:
Add tests, methods for groups

LivInTheLookingGlass committed:
add more tests, note that some are in progress

bcongdon commented:
So far the implementation LGTM. Looking forward to merging this 👍

LivInTheLookingGlass committed:
Add last resolution test

LivInTheLookingGlass committed:
Flake8 errors fixed

LivInTheLookingGlass commented:
Okay, I think everything I added is good. The only question I have left is if you'd like me to add more of my math utilities to this library instead. I moved the most useful already, but check this file for the rest: https://github.com/LivInTheLookingGlass/ManifoldMarketManager/blob/master/src/util.py

LivInTheLookingGlass committed:
Docstrings galore!

LivInTheLookingGlass committed:
More docstrings

LivInTheLookingGlass committed:
Pull out the nested functions, since they don't share scope

LivInTheLookingGlass commented:
Everything on  my end is ready. I also have an implementation of the comment API ready to go, but I haven't committed it because the endpoint is returning 401 errors

LivInTheLookingGlass committed:
Add create_comment()
Note that it is disabled at the moment because Manifold returns a 401 error

LivInTheLookingGlass committed:
Add some missing fields to Bet

LivInTheLookingGlass committed:
Enable comment API

LivInTheLookingGlass commented:
Done

bcongdon commented on pymanifold/lib.py:
```diff
@@ -119,35 +229,118 @@ def _create_market(
         )
+        if response.status_code in range(400, 500):
+            response.raise_for_status()
+        elif response.status_code >= 500:
```
Can you add a comment here why if there's an internal error we make a different API call?

bcongdon commented:
LGTM! just had one nitpick comment

LivInTheLookingGlass committed:
Add comment

LivInTheLookingGlass commented:
Just a heads up, I accidentally also pushed the comment API. It works and is stable, and has tests, but I wanted to make that a separate PR. Whoops

LivInTheLookingGlass committed:
Recursive types!

LivInTheLookingGlass committed:
Update typing scheme

bcongdon committed:
Update poetry

bcongdon commented:
Looks like all the tests are passing!
Thanks again for these changes, this is awesome 🎉
